DT: missing phandle-array entries fix

mbolivar-nordic opened the pull request:
Fixes: #30846 

The DTS language permits zeroing out phandles in a phandle array to say "there's nothing at this index", and dtlib manages that correctly, but edtlib and gen_defines.py aren't equipped to do so.

Fix this by allowing None elements in the lists of `ControllerAndData` values returned by edtlib for such properties.

Handle that in gen_defines.py by setting the generated `DT_N_<node>_P_<prop>_IDX_<i>_EXISTS` macro to 0 in such cases.

The `DT_N_<node>_P_<prop>_LEN` macro still accounts for the entire length of the phandle-array; it's just that some indexes may be missing data.

This series starts with a couple of cleanup commits.

mbolivar-nordic committed:
cmake: dts: remove stale error

We're well past Zephyr v1.16 (which turned out to be v2.1,
since v1.14 was followed by v2.0), so remove a message about a
variable that's been obsolete for two versions before that.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
dts: handle unspecified phandle-array elements

The DTS language permits zeroing out phandles in a phandle array to
say "there's nothing at this index", and dtlib manages that correctly,
but edtlib and gen_defines.py aren't equipped to do so.

Fix this by allowing None elements in the lists of ControllerAndData
values returned by edtlib for such properties.

Handle that in gen_defines.py by setting the generated
DT_N_<node>_P_<prop>_IDX_<i>_EXISTS macro to 0 in such cases.

The DT_N_<node>_P_<prop>_LEN macro still accounts for the entire
length of the phandle-array; it's just that some indexes may be
missing data.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

pabigot reviewed (approved):
Thanks, I think this will be very useful.

anangl commented on tests/lib/devicetree/api/src/main.c:
```diff
@@ -702,15 +702,25 @@ static void test_phandles(void)
 	zassert_false(DT_PHA_HAS_CELL_AT_IDX(TEST_PH, pha_gpios, 1, flags),
-		      "no pha-gpio 1 flags");
+		      "has pha-gpio 0 flags");
+	/* index 2 only has pin, no flags */
 	zassert_true(DT_PHA_HAS_CELL_AT_IDX(TEST_PH, pha_gpios, 2, pin),
+		      "no pha-gpio 1 pin");
+	zassert_false(DT_PHA_HAS_CELL_AT_IDX(TEST_PH, pha_gpios, 2, flags),
+		      "no pha-gpio 1 flags");
```
Several messages in the assertions here are quite confusing and misleading. For example, the above two should refer to index 2 and one of them should say "has" instead of "no", but actually I'm not sure which one, as I'm not sure if those messages are intended to say what was expected or what turned out to be incorrect.

mbolivar-nordic replied:
> Several messages in the assertions here are quite confusing and misleading.

You know, you're totally right. I think most of the assertions in this file are easy enough to read on their own, so I've added a cleanup commit at the beginning of the series that removes the ones that add nothing, or worse are actively confusing.

Thanks.

mbolivar-nordic commented on scripts/dts/edtlib.py:
```diff
@@ -2129,11 +2138,15 @@ def _add_names(node, names_ident, objs):
 
-        for obj, name in zip(objs, names):
+        for item in zip(objs, names):
+            if item is None:
```
Hm, I think this change is wrong, actually. I will add a test case covering this.

mbolivar-nordic replied:
>  I think this change is wrong, actually. I will add a test case covering this.

fixed

mbolivar-nordic committed:
dts: tests: delete extraneous whitespace

This line is required by the commit linter.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
tests: devicetree: remove confusing/useless strings

Many of the messages that are printed on failed test assertions in the
devicetre API tests provide little value (e.g. merely restating some
expression in the test) or are outright confusing (do they print the
thing that should have happened, or what went wrong?).

The ztest framework already prints the expression that failed, so
there's no need for this.

There are some exceptions, like in test_dep_ord(), where the messages
do print useful information, namely failing array indexes.

Keep those, but give the file a cleanup by removing the cruft. There
are alread cases where we use "" or NULL for self-explanatory
assertions in this file; we're just expanding that practice.

No changes to test results expected.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>
